addons/purchase/models/purchase.py

```diff
@@ -398,35 +398,104 @@ def _add_supplier_to_product(self):
                 except AccessError:  # no write access rights -> just ignore
                     break
 
+    def action_create_invoice(self):
+        self.ensure_one()
+        if self.invoice_status != 'to invoice':
+            raise UserError(_('There is no invoiceable line. If a product has a control policy based on received quantity, please make sure that a quantity has been received.'))
+
+        precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
+
+        # 1) Create invoice.
+        invoice_vals_list = []
+        pending_section = None
+        # Invoice values.
+        invoice_vals = self._prepare_invoice()
+        # Invoice line values (keep only necessary sections).
+        for line in self.order_line:
+            if line.display_type == 'line_section':
+                pending_section = line
+                continue
+            if not float_is_zero(line.qty_to_invoice, precision_digits=precision):
+                if pending_section:
+                    invoice_vals['invoice_line_ids'].append((0, 0, pending_section._prepare_account_move_line()))
+                    pending_section = None
+                invoice_vals['invoice_line_ids'].append((0, 0, line._prepare_account_move_line()))
+        invoice_vals_list.append(invoice_vals)
+
+        # 2) Create invoice.
+        move = self.env['account.move'].with_context(default_move_type='in_invoice').create(invoice_vals_list)
+        # 3) Some moves might actually be refunds: convert them if the total amount is negative
+        # We do this after the moves have been created since we need taxes, etc. to know if the total
+        # is actually negative or not
+        if move.amount_total >= 0:
+            action = self.env.ref('account.action_move_in_invoice_type').read()[0]
+        else:
+            action = self.env.ref('account.action_move_in_refund_type').read()[0]
+            move.action_switch_invoice_into_refund_credit_note()
+
+        action.update({
+            'views': [[False, 'form']],
+            'res_id': move.id,
+            'view_mode': 'form'
+        })
+        return action
+
+    def _prepare_invoice(self):
+        """
+        Prepare the dict of values to create the new invoice for a sales order. This method may be
+        overridden to implement custom invoice generation (making sure to call super() to establish
+        a clean extension chain).
+        """
+        self.ensure_one()
+        self = self.with_company(self.company_id)
+        journal = self.env['account.move'].with_company(self.company_id).with_context(default_type='in_invoice')._get_default_journal()
```

Comment:
@amoyaux I found we can't directly pass the journal_id. Because after doing that, every new account.move.line will have a account_id when are being created. Then in here
[https://github.com/odoo-dev/odoo/blob/31ed16d905370aadce6a00805b731d146d8e91cb/addons/account/models/account_move.py#L1510](url)
we can't go into this branch and use _get_computed_account() to get the correct account. Some module like stock_account override this method to change the account we want to use. 

My current solution is passing the company_id to the create values. It won't directly help us find a correct journal, but when create the account_move, we can use it to restrict the company we want to use:
[https://github.com/odoo-dev/odoo/blob/31ed16d905370aadce6a00805b731d146d8e91cb/addons/purchase/models/purchase.py#L464](url)